engine/context/src/main/java/io/deephaven/engine/context/QueryCompiler.java

```diff
@@ -488,7 +495,12 @@ private Class<?> compileHelper(@NotNull final String className,
                 knownClasses.put(identifyingFieldValue, p);
                 // It's also possible that some other code has already fulfilled this promise with exactly the same
                 // class. That's ok though: the promise code does not reject multiple sets to the identical value.
-                p.complete(result);
+                // But if the result is already set, that means that we've already been through this logic, determined
+                // the class, and set it (each run through this logic can be using a different ClassLoader, so it may
+                // be a different instance of the same Class).
+                if (!p.complete(result)) {
```

Comment:
We need to have a bit of a think about something, here. We "accidentally" fixed the issue Al reported when we switched to `CompletableFuture`, since it won't let us override the result. That said, that invalidates Corey's original commentary about `identifyingFieldValue`. If we will never overwrite the class in the future, we should also not be changing the key in `knownClasses`.

We should switch from `remove`+`put` to `get` followed by `remove` + `put` if and only if we succeed in `complete`.
That is, I think the code should be changed to this:
```
                CompletableFuture<Class<?>> future = knownClasses.get(identifyingFieldValue);
                final boolean found = future != null;
                if (!found) {
                    future = new CompletableFuture<>();
                }
                final boolean completed = future.complete(result);
                if (completed) {
                    if (found) {
                        knownClasses.remove(identifyingFieldValue);
                    }
                    knownClasses.put(identifyingFieldValue, future);
                }
```